src/utransport.rs

```diff
@@ -12,13 +12,112 @@
 
 use async_trait::async_trait;
+use std::hash::{Hash, Hasher};
+use std::ops::Deref;
+use std::sync::Arc;
 
 use crate::{UMessage, UStatus, UUri};
 
-/// `UTransport` is the uP-L1 interface that provides a common API for uE developers to send and receive messages.
+/// `UListener` is the uP-L1 interface that provides a means to create listeners which are registered to [`UTransport`]
 ///
-/// Implementations of `UTransport` contain the details for connecting to the underlying transport technology and
-/// sending `UMessage` using the configured technology. For more information, please refer to
+/// Implementations of `UListener` contain the details for what should occur when a message is received
+///
+/// For more information, please refer to
+/// [uProtocol Specification](https://github.com/eclipse-uprotocol/uprotocol-spec/blob/main/up-l1/README.adoc).
+///
+/// # Examples
+///
+/// ## Simple example
+///
+/// ```
+/// use up_rust::{UListener, UMessage, UStatus};
+///
+/// use async_trait::async_trait;
+/// use std::sync::{Arc, Mutex};
+///
+/// #[derive(Clone)]
+/// struct FooListener {
+///     inner_foo: Arc<Mutex<String>>
+/// }
+///
+/// #[async_trait]
+/// impl UListener for FooListener {
+///     async fn on_receive(&self, msg: UMessage) {
+///         let mut inner_foo = self.inner_foo.lock().unwrap();
+///         if let Some(payload) = msg.payload.as_ref() {
+///             if let Some(length) = payload.length.as_ref() {
+///                 *inner_foo = format!("latest message length: {length}");
+///             }
+///         }
+///     }
+///
+///     async fn on_error(&self, err: UStatus) {
+///         println!("uh oh, we got an error: {err:?}");
+///     }
+/// }
+/// ```
+///
+/// ## Long-running function needed when message received
+///
+/// ```
+/// use up_rust::{UListener, UMessage, UStatus};
+///
+/// use async_trait::async_trait;
+/// use async_std::task;
+/// use std::sync::{Arc, Mutex};
+///
+/// #[derive(Clone)]
+/// struct LongTaskListener;
+///
+/// async fn send_to_jupiter(message_for_jupiter: UMessage) {
+///     // send a message to Jupiter
+///     println!("Fly me to the moon... {message_for_jupiter}");
+/// }
+///
+/// #[async_trait]
+/// impl UListener for LongTaskListener {
+///     async fn on_receive(&self, msg: UMessage) {
+///         task::spawn(send_to_jupiter(msg));
+///     }
+///
+///     async fn on_error(&self, err: UStatus) {
+///         println!("unable to send to jupiter :( {err:?}");
+///     }
+/// }
+/// ```
+#[async_trait]
+pub trait UListener: 'static + Send + Sync {
+    /// Performs some action on receipt of a message
+    ///
+    /// # Parameters
+    ///
+    /// * `msg` - The message
+    ///
+    /// # Note
+    ///
+    /// `on_receive()` is expected to return almost immediately. If it does not, it could potentially
+    /// block further message receipt. For long-running, synchronous, or blocking operations consider
```

Comment:
Ahh, you would like a mention of this for implementers of `UTransport`?

I was writing this for the audience of "writers of uEs".

I suppose an implementation of `UTransport` could do one of two things:
1. `await` on `on_receive()`
2. spawn a new async task to run `on_receive()`

In the case of 1, it would matter that `on_recieve()` return fast.

In the case of 2, it becomes a bit more of a gray area on whether it must return fast and may depend on how the `on_receive()` manages any shared mutable state which would prevent forward progress on any future calls to `on_receive()`.

Re-reading what you wrote now, perhaps your point is that since a `UTransport` may do 2, the UListener must be able to be called again immediately for the next incoming message.

Almost seems like to me having two sections of caveats, one for implementers of UListener and another for implementers of UTransport may be helpful.

WDYT?